File: data/harnesses/Plus/cpp/secure/CWE-434_09/main.cpp
#include <string>
#include <vector>
#include <filesystem>
#include <fstream>
#include <regex>
#include <stdexcept>
#include <algorithm>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

std::string save_profile_document(const std::string& user_id, 
                                  const std::string& filename, 
                                  const std::vector<unsigned char>& content_bytes, 
                                  const std::string& upload_dir) {
    if (user_id.empty() || user_id.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("empty user_id");
    }
    if (filename.empty() || filename.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("empty filename");
    }
    if (upload_dir.empty() || upload_dir.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("empty upload_dir");
    }

    std::regex user_id_regex("^[A-Za-z0-9_-]+$");
    if (!std::regex_match(user_id, user_id_regex)) {
        throw std::invalid_argument("invalid user_id");
    }

    std::string lower_filename = filename;
    std::transform(lower_filename.begin(), lower_filename.end(), lower_filename.begin(),
                   [](unsigned char c){ return std::tolower(c); });
    if (lower_filename.size() < 4 || lower_filename.substr(lower_filename.size() - 4) != ".pdf") {
        throw std::invalid_argument("unsupported extension");
    }

    if (content_bytes.size() > 1024 * 1024) {
        throw std::invalid_argument("file too large");
    }

    fs::create_directories(upload_dir);

    fs::path safe_path = fs::path(upload_dir) / (user_id + "_profile.pdf");
    std::ofstream out_file(safe_path, std::ios::binary);
    if (!out_file) {
        throw std::runtime_error("failed to open file for writing");
    }
    if (!content_bytes.empty()) {
        out_file.write(reinterpret_cast<const char*>(content_bytes.data()), content_bytes.size());
    }
    out_file.close();

    return safe_path.string();
}

// Test helpers
int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " - Expected '" + expected + "' but got '" + actual + "'");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, still counts as raised
    }
}

int main() {
    std::string base = "./Test/CWE-434_09";
    
    // Clean up and create base directory
    std::error_code ec;
    fs::remove_all(base, ec);
    fs::create_directories(base, ec);

    // 1) Valid pdf save ignores original name and uses canonical name
    std::vector<unsigned char> content1 = {'%', 'P', 'D', 'F', '-', '1', '.', '4', '\n', '.', '.', '.'};
    std::string out1 = save_profile_document("user_1", "resume.pdf", content1, base);
    expect_equal(out1, (fs::path(base) / "user_1_profile.pdf").string(), "Test 1 path");
    expect_true(fs::is_regular_file(out1), "Test 1 file exists");

    // 2) Original filename with path traversal but .pdf should still be accepted, output stays canonical
    std::vector<unsigned char> content2 = {'%', 'P', 'D', 'F', '-', '1', '.', '4', '\n', 'x'};
    std::string out2 = save_profile_document("user-2", "../evil.pdf", content2, base);
    expect_equal(out2, (fs::path(base) / "user-2_profile.pdf").string(), "Test 2 path");
    expect_true(fs::is_regular_file(out2), "Test 2 file exists");

    // 3) Disallowed extension should raise
    std::vector<unsigned char> content3_exe = {'M', 'Z', '.', '.', '.'};
    assert_raises([&]() {
        save_profile_document("user3", "script.exe", content3_exe, base);
    }, "Test 3a unsupported extension .exe");

    std::vector<unsigned char> content3_txt = {'h', 'i'};
    assert_raises([&]() {
        save_profile_document("user3", "note.txt", content3_txt, base);
    }, "Test 3b unsupported extension .txt");

    // 4) Too large should raise
    std::vector<unsigned char> content4(1024 * 1024 + 1, 'x');
    assert_raises([&]() {
        save_profile_document("user4", "big.pdf", content4, base);
    }, "Test 4 file too large");

    // 5) Invalid user_id should raise
    std::vector<unsigned char> content5 = {'%', 'P', 'D', 'F'};
    assert_raises([&]() {
        save_profile_document("user 5", "a.pdf", content5, base);
    }, "Test 5a invalid user_id (space)");

    assert_raises([&]() {
        save_profile_document("", "a.pdf", content5, base);
    }, "Test 5b invalid user_id (empty)");

    // 6) Wrong types simulated (C++ is statically typed, so we simulate Python TypeError equivalents)
    // 6a: Empty user_id simulating None/invalid type input
    assert_raises([&]() {
        save_profile_document("", "a.pdf", content5, base);
    }, "Test 6a invalid user_id (simulated wrong type)");

    // 6b: Empty filename simulating None/invalid type input
    assert_raises([&]() {
        save_profile_document("user6", "", content5, base);
    }, "Test 6b invalid filename (simulated wrong type)");

    // 6c: Empty upload_dir simulating None/invalid type input
    assert_raises([&]() {
        save_profile_document("user6", "a.pdf", content5, "");
    }, "Test 6c invalid upload_dir (simulated wrong type)");

    // Cleanup
    fs::remove_all(base, ec);

    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-434_09/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)
var userIDRegex = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

func saveProfileDocument(userID string, filename string, contentBytes []byte, uploadDir string) (string, error) {
	if strings.TrimSpace(userID) == "" {
		return "", errors.New("empty user_id")
	}
	if !userIDRegex.MatchString(userID) {
		return "", errors.New("invalid user_id")
	}
	if strings.TrimSpace(filename) == "" {
		return "", errors.New("empty filename")
	}
	if !strings.HasSuffix(strings.ToLower(filename), ".pdf") {
		return "", errors.New("unsupported extension")
	}
	if len(contentBytes) > 1024*1024 {
		return "", errors.New("file too large")
	}
	if strings.TrimSpace(uploadDir) == "" {
		return "", errors.New("empty upload_dir")
	}

	err := os.MkdirAll(uploadDir, 0755)
	if err != nil {
		return "", fmt.Errorf("failed to create upload directory: %w", err)
	}

	safePath := filepath.Join(uploadDir, fmt.Sprintf("%s_profile.pdf", userID))

	err = os.WriteFile(safePath, contentBytes, 0644)
	if err != nil {
		return "", fmt.Errorf("failed to write file: %w", err)
	}

	return safePath, nil
}

func main() {
	base := filepath.Join(".", "Test", "CWE-434_09")
	os.RemoveAll(base)
	os.MkdirAll(base, 0755)

	passed := 0
	failed := 0

	assertEqual := func(name string, got, expected interface{}) {
		if got != expected {
			fmt.Printf("FAIL %s: got %v, expected %v\n", name, got, expected)
			failed++
		} else {
			passed++
		}
	}

	assertError := func(name string, err error) {
		if err == nil {
			fmt.Printf("FAIL %s: expected error but got nil\n", name)
			failed++
		} else {
			passed++
		}
	}

	// 1) valid pdf save ignores original name and uses canonical name
	out1, err1 := saveProfileDocument("user_1", "resume.pdf", []byte("%PDF-1.4\n..."), base)
	if err1 != nil {
		fmt.Printf("FAIL test 1: unexpected error: %v\n", err1)
		failed++
	} else {
		expected1 := filepath.Join(base, "user_1_profile.pdf")
		assertEqual("test 1 path", out1, expected1)
		info1, statErr := os.Stat(out1)
		if statErr != nil {
			fmt.Printf("FAIL test 1 stat: %v\n", statErr)
			failed++
		} else if info1.IsDir() {
			fmt.Println("FAIL test 1: expected file, got dir")
			failed++
		} else {
			passed++
		}
	}

	// 2) original filename with path traversal but .pdf should still be accepted, output stays canonical
	out2, err2 := saveProfileDocument("user-2", "../evil.pdf", []byte("%PDF-1.4\nx"), base)
	if err2 != nil {
		fmt.Printf("FAIL test 2: unexpected error: %v\n", err2)
		failed++
	} else {
		expected2 := filepath.Join(base, "user-2_profile.pdf")
		assertEqual("test 2 path", out2, expected2)
		info2, statErr := os.Stat(out2)
		if statErr != nil {
			fmt.Printf("FAIL test 2 stat: %v\n", statErr)
			failed++
		} else if info2.IsDir() {
			fmt.Println("FAIL test 2: expected file, got dir")
			failed++
		} else {
			passed++
		}
	}

	// 3) disallowed extension should raise
	_, err3a := saveProfileDocument("user3", "script.exe", []byte("MZ..."), base)
	assertError("test 3a unsupported extension", err3a)

	_, err3b := saveProfileDocument("user3", "note.txt", []byte("hi"), base)
	assertError("test 3b unsupported extension", err3b)

	// 4) too large should raise
	_, err4 := saveProfileDocument("user4", "big.pdf", make([]byte, 1024*1024+1), base)
	assertError("test 4 file too large", err4)

	// 5) invalid user_id should raise
	_, err5a := saveProfileDocument("user 5", "a.pdf", []byte("%PDF"), base)
	assertError("test 5a invalid user_id", err5a)

	_, err5b := saveProfileDocument("", "a.pdf", []byte("%PDF"), base)
	assertError("test 5b empty user_id", err5b)

	// 6) empty filename should raise
	_, err6a := saveProfileDocument("user6", "", []byte("%PDF"), base)
	assertError("test 6a empty filename", err6a)

	// 7) empty upload_dir should raise
	_, err7 := saveProfileDocument("user7", "a.pdf", []byte("%PDF"), "")
	assertError("test 7 empty upload_dir", err7)

	// Cleanup
	os.RemoveAll(base)

	fmt.Printf("\n%d passed, %d failed\n", passed, failed)
	if failed > 0 {
		os.Exit(1)
	}
}